### json 불러와서 저장하기

In [39]:
import numpy as np
import pandas as pd
import json
import jpype
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

data_path = "data"
#data_path = "data(test)"

file_list = glob.glob("%s/*.json" % data_path)

json_train=[]

shuffle(file_list)
for json_file_name in file_list:
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]

### json에서 author, author 형태소, forumid 추출

In [40]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    if cnt % 10000 == 0:
        print cnt

    labeled_train.append({
            "istroll": article["is_troll"],
            "author": article["author"],
            "author_pos": ["%s_%s" % (word, pos) for word, pos in mecab.pos(article["author"])],
            "forumid": article["forumid"],
            "pk": article["pk"]
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train = labeled_train.set_index('pk')

labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


author                                         author_pos  \
pk                                                                              
64993                메론바흐규                            [메론_NNP, 바흐_NNP, 규_NNG]   
64994                메론바흐규                            [메론_NNP, 바흐_NNP, 규_NNG]   
64995              Eliciel                                       [Eliciel_SL]   
64996                  김카디                                    [김_NNP, 카디_NNG]   
64997                 한줌의흙                             [한줌_NNG, 의_JKG, 흙_NNG]   
64998                   라트                                           [라트_NNP]   
64999         민영화는대체어떤냔이얌?  [민영화_NNG, 는_JX, 대체_MAG, 어떤_MM, 냔_ETM, 이얌_NNG, ...   
65000                 LEOD                                          [LEOD_SL]   
65001              Eliciel                                       [Eliciel_SL]   
65002           이모춘천다까끼추가요  [이모_NNG, 춘천_NNG, 다_MAG, 까끼_NNG, 추가_NNG, 요_VCP+EC]   
65003               044APD                                   [044_SN, APD_SL]   
65004                일베사냥꾼                       [일_NR, 베_NNG, 사냥_NNG, 꾼_XSN]   
65005              호식이2마리반              [호식_NNG, 이_JKS, 2_SN, 마리_NNBC, 반_NNG]   
65006                  김탁구                                    [김_NNP, 탁구_NNG]   
65007                 질소까스                                   [질소_NNG, 까스_NNP]   
65008                hakun                                         [hakun_SL]   
65009                  하현달                                          [하현달_NNG]   
65010              벌레는족쳐야지                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]   
65011                 댓통박령                              [댓_NR, 통박_NNG, 령_NNG]   
65012              Eliciel                                       [Eliciel_SL]   
65013                  드롯셀                              [드_NNP, 롯_NNP, 셀_NNG]   
65014                  레민준                                    [레_NNP, 민준_NNP]   
65015              벌레는족쳐야지                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]   
65016               밭정보수집중                     [밭_NNG, 정보_NNG, 수집_NNG, 중_NNB]   
65017                  회계왕                                    [회계_NNG, 왕_NNG]   
65018                  돌핀스                                          [돌핀스_NNP]   
65019              Eliciel                                       [Eliciel_SL]   
65020                닭!잡는닭                  [닭_NNG, !_SF, 잡_VV, 는_ETM, 닭_NNG]   
65021                  돌핀스                                          [돌핀스_NNP]   
65022                   서백                                           [서백_NNP]   
...                    ...                                                ...   
189619         아이라인 그리는 판다                   [아이라인_NNG, 그리_VV, 는_ETM, 판다_NNG]   
189620                 박아희                                    [박_NNP, 아희_NNG]   
189621      꾸꾸까까 ʕ·͡ᴥ·ʔ/ ♥  [꾸꾸_MAG, 까까_NNG, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC, ʔ_SY...   
189623                  정부                                           [정부_NNG]   
189624                  정부                                           [정부_NNG]   
189625                  정부                                           [정부_NNG]   
189626                  정부                                           [정부_NNG]   
189627                  정부                                           [정부_NNG]   
189628                酒袋飯囊                              [酒_XSN, 袋_SH, 飯囊_NNG]   
189629                미식헌터                         [미식_NNG, 헌_XSV+ETM, 터_NNB]   
189630            휴가중인 보도국                [휴가_NNG, 중_NNB, 인_VCP+ETM, 보도국_NNG]   
189631                  정부                                           [정부_NNG]   
189632                  정부                                           [정부_NNG]   
189633                뷰3유저                              [뷰_NNP, 3_SN, 유저_NNG]   
189634    방해:AVI ʕ·͡ᴥ·ʔ/ ♥  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...   
189635                 모히토                                     [모_IC, 히토_NNP]   
189636    방해:AVI ʕ·͡ᴥ·ʔ/ ♥  [방해_NNG, :_SC, AVI_SL, 

### forumid와 author을 label encoder로 숫자로 바꿔줌

In [41]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()

labeled_train["forumid"] = le.fit_transform(labeled_train["forumid"])
labeled_train["author"] = le.fit_transform(labeled_train["author"])

labeled_train

author                                         author_pos  forumid  \
pk                                                                           
64993    12009                            [메론_NNP, 바흐_NNP, 규_NNG]       25   
64994    12009                            [메론_NNP, 바흐_NNP, 규_NNG]       14   
64995     1011                                       [Eliciel_SL]       31   
64996     6770                                    [김_NNP, 카디_NNG]       31   
64997    25906                             [한줌_NNG, 의_JKG, 흙_NNG]        1   
64998    10909                                           [라트_NNP]       29   
64999    12784  [민영화_NNG, 는_JX, 대체_MAG, 어떤_MM, 냔_ETM, 이얌_NNG, ...       31   
65000     1394                                          [LEOD_SL]       31   
65001     1011                                       [Eliciel_SL]       31   
65002    20490  [이모_NNG, 춘천_NNG, 다_MAG, 까끼_NNG, 추가_NNG, 요_VCP+EC]       19   
65003       78                                   [044_SN, APD_SL]       17   
65004    21017                       [일_NR, 베_NNG, 사냥_NNG, 꾼_XSN]       27   
65005    26387              [호식_NNG, 이_JKS, 2_SN, 마리_NNBC, 반_NNG]       32   
65006     6771                                    [김_NNP, 탁구_NNG]       27   
65007    22925                                   [질소_NNG, 까스_NNP]       31   
65008     2813                                         [hakun_SL]       12   
65009    25794                                          [하현달_NNG]       31   
65010    13784                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]       31   
65011    10000                              [댓_NR, 통박_NNG, 령_NNG]       27   
65012     1011                                       [Eliciel_SL]       12   
65013    10543                              [드_NNP, 롯_NNP, 셀_NNG]       12   
65014    11017                                    [레_NNP, 민준_NNP]       31   
65015    13784                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]       32   
65016    13501                     [밭_NNG, 정보_NNG, 수집_NNG, 중_NNB]       32   
65017    26594                                    [회계_NNG, 왕_NNG]       32   
65018    10317                                          [돌핀스_NNP]       32   
65019     1011                                       [Eliciel_SL]       32   
65020     9310                  [닭_NNG, !_SF, 잡_VV, 는_ETM, 닭_NNG]       32   
65021    10317                                          [돌핀스_NNP]       32   
65022    15629                                           [서백_NNP]       31   
...        ...                                                ...      ...   
189619   17442                   [아이라인_NNG, 그리_VV, 는_ETM, 판다_NNG]       31   
189620   13230                                    [박_NNP, 아희_NNG]       34   
189621    7078  [꾸꾸_MAG, 까까_NNG, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC, ʔ_SY...       14   
189623   22011                                           [정부_NNG]       12   
189624   22011                                           [정부_NNG]       31   
189625   22011                                           [정부_NNG]       31   
189626   22011                                           [정부_NNG]       31   
189627   22011                                           [정부_NNG]        4   
189628    4591                              [酒_XSN, 袋_SH, 飯囊_NNG]       31   
189629   12681                         [미식_NNG, 헌_XSV+ETM, 터_NNB]       31   
189630   26685                [휴가_NNG, 중_NNB, 인_VCP+ETM, 보도국_NNG]       31   
189631   22011                                           [정부_NNG]       31   
189632   22011                                           [정부_NNG]        5   
189633   14581                              [뷰_NNP, 3_SN, 유저_NNG]        1   
189634   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
189635   12213                                     [모_IC, 히토_NNP]        2   
189636   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
189637     209  [1_SN, 붸뷉충아_UNKNOWN, 아프_VA, 지_EC, 말_VX, 고_EC, ...       31   
189638   12648                

###bag of words

In [42]:
from sklearn.feature_extraction.text import CountVectorizer

def make_bag_of_words(labeled_train, max_features, col_name):
    vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                                stop_words = None, max_features=max_features)
    
    sentences = labeled_train[col_name].apply(lambda x:" ".join(x))
    train_data_features = vectorizer.fit_transform(sentences).toarray()

    col = ["bow_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    #col = ["bow%s%s" % (col_name, str(i)) for  in range(0, max_features)]
    df_bow = pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)
    
    labeled_train = pd.concat([labeled_train, df_bow],axis=1)
    
    return labeled_train

##LDA

In [43]:
import gensim
from gensim import corpora, models
from gensim.corpora import TextCorpus, MmCorpus, Dictionary
from nltk.corpus import stopwords
from optparse import OptionParser

def make_lda(train, keep_n, num_topics, col_name):

    data = train[col_name]

    dictionary = corpora.Dictionary(data)
    dictionary.filter_extremes(keep_n=keep_n)

    corpus = [dictionary.doc2bow(text) for text in data]

    print("Make Lda..")

#    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1,distributed = True)
    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1)
    
    num = len(train)    
    df = []
    
    for i in range(0,num):
        if i % 10000 == 0:
            print(i)
        
        temp = [i[1] for i in lda.get_document_topics(corpus[i],minimum_probability=0)]
        #temp = max(lda.get_document_topics(corpus[i]),key=lambda item:item[1])[0]
        df.append(temp)
    
    col = ["lda_%s_%d" % (col_name, data) for data in range(0, num_topics)]
    df = pd.DataFrame(df, columns = col)
    df.index = train.index
    
    train = pd.concat([train, df], axis=1)
        
    return train

##Word2Vec

In [45]:
from gensim.models import word2vec

def makeFeatureVec(words, model, num_features):
    
    featureVec = np.zeros((num_features,), dtype = "float32")
    
    nwords = 0
    
    index2word_set = set(model.index2word)
    
    for word in words:
        if word in index2word_set:
            nwords = nwords + 1.
            featureVec = np.add(featureVec, model[word])
    
    featureVec = np.divide(featureVec, nwords)
    
    return featureVec

def getAvgFeatureVecs(texts, model, num_features):
    
    counter = 0
    
    textFeatureVecs = np.zeros((len(texts), num_features), dtype = "float32")
    
    for i, text in enumerate(texts):
        
        if i % 10000 == 0:
            print(i)
        
        textFeatureVecs[i] = makeFeatureVec(text , model, num_features)
        
    return textFeatureVecs

def make_word2vec(train, col_name):
    num_features = 300
    min_word_count = 40
    num_workers = 4
    context = 10
    downsampling = 1e-3
    
    sentences = " ".join(train[col_name].apply(lambda x:" ".join(x)))
    
    model = word2vec.Word2Vec(sentences, workers = num_workers, size = num_features,\
                             min_count = min_word_count, window = context, sample = downsampling)
    
    train_feature = getAvgFeatureVecs(train[col_name].apply(lambda x:" ".join(x)), model, num_features)
    train_feature = pd.DataFrame(train_feature, index = train.index)
    
    train = pd.concat([train, train_feature], axis = 1)
    
    return train

In [46]:
max_features = 1000
# #%time labeled_train = make_lda(labeled_train, 1000, 200, "author_pos")
# #%time labeled_train = make_bag_of_words(labeled_train, 1000, "author_pos")

%time labeled_train = make_word2vec(labeled_train, "author_pos")
labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000
CPU times: user 36.6 s, sys: 12.1 s, total: 48.8 s
Wall time: 40.3 s


author                                         author_pos  forumid  \
pk                                                                           
64993    12009                            [메론_NNP, 바흐_NNP, 규_NNG]       25   
64994    12009                            [메론_NNP, 바흐_NNP, 규_NNG]       14   
64995     1011                                       [Eliciel_SL]       31   
64996     6770                                    [김_NNP, 카디_NNG]       31   
64997    25906                             [한줌_NNG, 의_JKG, 흙_NNG]        1   
64998    10909                                           [라트_NNP]       29   
64999    12784  [민영화_NNG, 는_JX, 대체_MAG, 어떤_MM, 냔_ETM, 이얌_NNG, ...       31   
65000     1394                                          [LEOD_SL]       31   
65001     1011                                       [Eliciel_SL]       31   
65002    20490  [이모_NNG, 춘천_NNG, 다_MAG, 까끼_NNG, 추가_NNG, 요_VCP+EC]       19   
65003       78                                   [044_SN, APD_SL]       17   
65004    21017                       [일_NR, 베_NNG, 사냥_NNG, 꾼_XSN]       27   
65005    26387              [호식_NNG, 이_JKS, 2_SN, 마리_NNBC, 반_NNG]       32   
65006     6771                                    [김_NNP, 탁구_NNG]       27   
65007    22925                                   [질소_NNG, 까스_NNP]       31   
65008     2813                                         [hakun_SL]       12   
65009    25794                                          [하현달_NNG]       31   
65010    13784                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]       31   
65011    10000                              [댓_NR, 통박_NNG, 령_NNG]       27   
65012     1011                                       [Eliciel_SL]       12   
65013    10543                              [드_NNP, 롯_NNP, 셀_NNG]       12   
65014    11017                                    [레_NNP, 민준_NNP]       31   
65015    13784                    [벌레_NNG, 는_JX, 족쳐야_VV+EC, 지_VX]       32   
65016    13501                     [밭_NNG, 정보_NNG, 수집_NNG, 중_NNB]       32   
65017    26594                                    [회계_NNG, 왕_NNG]       32   
65018    10317                                          [돌핀스_NNP]       32   
65019     1011                                       [Eliciel_SL]       32   
65020     9310                  [닭_NNG, !_SF, 잡_VV, 는_ETM, 닭_NNG]       32   
65021    10317                                          [돌핀스_NNP]       32   
65022    15629                                           [서백_NNP]       31   
...        ...                                                ...      ...   
189619   17442                   [아이라인_NNG, 그리_VV, 는_ETM, 판다_NNG]       31   
189620   13230                                    [박_NNP, 아희_NNG]       34   
189621    7078  [꾸꾸_MAG, 까까_NNG, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC, ʔ_SY...       14   
189623   22011                                           [정부_NNG]       12   
189624   22011                                           [정부_NNG]       31   
189625   22011                                           [정부_NNG]       31   
189626   22011                                           [정부_NNG]       31   
189627   22011                                           [정부_NNG]        4   
189628    4591                              [酒_XSN, 袋_SH, 飯囊_NNG]       31   
189629   12681                         [미식_NNG, 헌_XSV+ETM, 터_NNB]       31   
189630   26685                [휴가_NNG, 중_NNB, 인_VCP+ETM, 보도국_NNG]       31   
189631   22011                                           [정부_NNG]       31   
189632   22011                                           [정부_NNG]        5   
189633   14581                              [뷰_NNP, 3_SN, 유저_NNG]        1   
189634   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
189635   12213                                     [모_IC, 히토_NNP]        2   
189636   13482  [방해_NNG, :_SC, AVI_SL, ʕ_SY, ·_SC, ͡ᴥ_SY, ·_SC...       31   
189637     209  [1_SN, 붸뷉충아_UNKNOWN, 아프_VA, 지_EC, 말_VX, 고_EC, ...       31   
189638   12648                

### predictor, model 생성

In [49]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['author_pos', 'author', label])
labeled_train = labeled_train.fillna(0)

model = RandomForestClassifier(n_estimators=10, n_jobs=6)

###cross validation

In [50]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")

cv_result = scores.mean()

print(cv_result)

0.869585615353


## Result

 * author를 Label Encoder로 변환: 0.67331271058
 * author의 morphs를 Bag of Words(feature = 1,000 개)로 변환: 0.733333689652
 * author의 pos를 Bag of Words(feature = 1,000 개)로 변환: 0.769833271386
 * 데이터를 shuffle하고 pos를 Bag of Words(feautre = 1,000개)로 변환: 0.81172740086
 
 
 * author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.748718319834
 * 데이터를 shuffle하고 author의 pos를 LDA로 변환 (num_topics = 200, keep_n = 1,000) : 0.772313052394
 
 * 데이터 shuffle 후 bag of words(num_features = 300) : 0.869585615353

In [51]:
import pickle

pre = labeled_train.columns.drop(['author_pos', 'author'])

pickle.dump(labeled_train[pre], open("author, forum_%d.p" % max_features, "wb"), protocol = pickle.HIGHEST_PROTOCOL)